x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformScheduler.java

```diff
@@ -0,0 +1,388 @@
+    public static class Event {
+        private final String transformId;
+        private final long scheduledTime;
+        private final long triggeredTime;
+
+        public Event(String transformId, long scheduledTime, long triggeredTime) {
+            this.transformId = transformId;
+            this.scheduledTime = scheduledTime;
+            this.triggeredTime = triggeredTime;
+        }
+
+        public String getTransformId() {
+            return transformId;
+        }
+
+        public long getScheduledTime() {
+            return scheduledTime;
+        }
+
+        public long getTriggeredTime() {
+            return triggeredTime;
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (this == other) return true;
+            if (other == null || getClass() != other.getClass()) return false;
+            Event that = (Event) other;
+            return Objects.equals(this.transformId, that.transformId)
+                && this.scheduledTime == that.scheduledTime
+                && this.triggeredTime == that.triggeredTime;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(transformId, scheduledTime, triggeredTime);
+        }
+
+        @Override
+        public String toString() {
+            return new StringBuilder("Event[").append("transformId=")
+                .append(transformId)
+                .append(",scheduledTime=")
+                .append(Instant.ofEpochMilli(scheduledTime))
+                .append(",triggeredTime=")
+                .append(Instant.ofEpochMilli(triggeredTime))
+                .append("]")
+                .toString();
+        }
+    }
+
+    /**
+     * {@link Listener} interface allows receiving events about the scheduled task being triggered.
+     */
+    public interface Listener {
+        /**
+         * Called whenever scheduled task is being triggered.
+         *
+         * @param event event structure containing data that might be useful to the listener
+         */
+        void triggered(Event event);
+    }
+
+    /**
+     * {@link ScheduledTransformTask} is a structure describing the scheduled task in the queue.
+     *
+     * This class is immutable.
+     */
+    static class ScheduledTransformTask {
+
+        /** Minimum delay that can be applied after a failure. */
+        private static final long MIN_DELAY_MILLIS = Duration.ofSeconds(1).toMillis();
+        /** Maximum delay that can be applied after a failure. */
+        private static final long MAX_DELAY_MILLIS = Duration.ofHours(1).toMillis();
+
+        private final String transformId;
+        private final TimeValue frequency;
+        private final Long lastTriggeredTimeMillis;
+        private final int failureCount;
+        private final long nextScheduledTimeMillis;
+        private final Listener listener;
+
+        ScheduledTransformTask(
+            String transformId,
+            TimeValue frequency,
```

Comment:
I wonder: could we use `SchedulerEngine.Schedule` here? This would allow e.g. adding new types of scheduling like cron-based?

(however we can push that out to whenever it gets implemented)